README.md
# Spantastic Library

Spantastic Library is a lightweight Android library for applying custom spans on text in TextView. It provides an easy way to apply custom styles, fonts, colors, and click events to specific parts of the text in a TextView.

Current Version: [![](https://jitpack.io/v/thesarangal/spantasticlibrary.svg)](https://jitpack.io/#thesarangal/spantasticlibrary)

## Features

- Apply custom spans to specific parts of the text in a TextView.
- Support for custom fonts, colors, styles, and click events.
- Lightweight and easy to use.

## How to To get a Git project into your build

### Step 1. Add the JitPack repository to your build file

Add it in your root build.gradle at the end of repositories:

	allprojects {
		repositories {
			...
			maven { url 'https://jitpack.io' }
		}
	}

### Step 2. Add the dependency

	dependencies {
	        implementation 'com.github.thesarangal:spantasticlibrary:VERSION_CODE'
	}

Done! The first time you request a project JitPack checks out the code, builds it and serves the build artifacts (jar, aar).

## How to use Spantastic
### I. Just Show Underline under specific word

    Java:

	String sentence = "Spannable with single word underline.";
	TextView tvOne = findViewById(R.id.tv_one);
	new Spantastic.Builder(tvOne, sentence)
                    .setSpan("underline")
                    .showUnderline(true)
                    .apply();


    Kotlin:

    var sentence = "Spannable with single word underline."
    val tvOne = findViewById<TextView>(R.id.tv_one)
    Spantastic.Builder(tvOne, sentence)
                    .setSpan("underline")
                    .showUnderline(true)
                    .apply()

		
### II. Show Underline and make Text BOLD for Multiple words

    Java

	String sentence = "Spannable with underline and bold for multiple words.";
	List<String> strings = new ArrayList<>();
	strings.add("spannable");
	strings.add("underline");
	strings.add("bold");
	TextView tvTwo = findViewById(R.id.tv_two);
	new Spantastic.Builder(tvTwo, sentence)
                    .setSpanList(strings)
                    .showUnderline(true)
                    .setTypeface(Typeface.create(Typeface.DEFAULT, Typeface.BOLD))
                    .apply();


    Kotlin

    sentence = "Spannable with underline and bold for multiple words."
    var strings: ArrayList<String> = ArrayList()
    strings.add("spannable")
    strings.add("underline")
    strings.add("bold")
    val tvTwo = findViewById<TextView>(R.id.tv_two)
    Spantastic.Builder(tvTwo, sentence)
                    .setSpanList(strings)
                    .showUnderline(true)
                    .setTypeface(Typeface.create(Typeface.DEFAULT, Typeface.BOLD))
                    .apply()

### III. Set Click on Multiple Words with Bold Style

    Java

	String sentence = "Spannable with click, color, underline and bold for multiple words.";
	List<String> strings = new ArrayList<>();
	strings.add("click");
	strings.add("color");
	strings.add("underline");
	strings.add("bold");
	TextView tvThree = findViewById(R.id.tv_three);
	new Spantastic.Builder(tvThree, sentence)
                    .setSpanList(strings)
                    .showUnderline(true)
                    .setSpanColor(ContextCompat.getColor(this, R.color.colorAccent))
                    .setTypeface(Typeface.create(Typeface.DEFAULT, Typeface.BOLD))
                    .setClickCallback(new Spantastic.SpannableCallBack() {
                        @Override
                        public void onSpanClick(String spanString, Object... object) {
                            Toast.makeText(MainActivity.this, String.format(Locale.ENGLISH,
                                    "\"%s\" Word Clicked", spanString), Toast.LENGTH_SHORT)
                                    .show();
                        }
                    })
                    .apply();


    Kotlin

    sentence =
        "Spannable with click, color, underline and bold for multiple words."
    strings = ArrayList()
    strings.add("click")
    strings.add("color")
    strings.add("underline")
    strings.add("bold")
    val tvThree = findViewById<TextView>(R.id.tv_three)
    Spantastic.Builder(tvThree, sentence)
                    .setSpanList(strings)
                    .showUnderline(true)
                    .setSpanColor(ContextCompat.getColor(this, R.color.colorAccent))
                    .setTypeface(Typeface.create(Typeface.DEFAULT, Typeface.BOLD))
                    .setClickCallback(object : Spantastic.SpannableCallBack {
                        override fun onSpanClick(spanString: String?, vararg `object`: Any?) {
                            Toast.makeText(
                                this@KotlinActivity, String.format(
                                    Locale.ENGLISH,
                                    "\"%s\" Word Clicked", spanString
                                ), Toast.LENGTH_SHORT
                            )
                                .show()
                        }
                    })
                    .apply()
		
### IV. Change TextSize

    Java

	String sentence = "Spantastic with custom text size for words.";
    TextView tvFour = findViewById(R.id.tv_four);
    new Spantastic.Builder(tvFour, sentence)
                        .setSpan("text size")
                        .setTextSize(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, 20,
                                this.getResources().getDisplayMetrics()))
                        .apply();


    Kotlin

    sentence = "Spantastic with custom text size for words."
    val tvFour = findViewById<TextView>(R.id.tv_four)
    Spantastic.Builder(tvFour, sentence)
                        .setSpan("text size")
                        .setTextSize(
                            TypedValue.applyDimension(
                                TypedValue.COMPLEX_UNIT_SP, 20f,
                                this.resources.displayMetrics
                            )
                        )
                        .apply()

		
### V. sentence = "Spantastic with custom styles for multiple words.";

    Java

	TextView tvFive = findViewById(R.id.tv_five);
	List<SpanModel> spanModelList = new ArrayList<>();
	spanModelList.add(new SpanModel("custom", ContextCompat.getColor(this, R.color.colorPrimary), "custom", true, Typeface.create(Typeface.DEFAULT, Typeface.NORMAL), TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, 15, this.getResources().getDisplayMetrics())));
	spanModelList.add(new SpanModel("styles", ContextCompat.getColor(this, R.color.colorPrimaryDark), "styles", false, Typeface.create(Typeface.DEFAULT, Typeface.BOLD), TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, 10, this.getResources().getDisplayMetrics())));
	spanModelList.add(new SpanModel("for", ContextCompat.getColor(this, R.color.colorAccent), "for", true, Typeface.create(Typeface.SERIF, Typeface.ITALIC), TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, 18, this.getResources().getDisplayMetrics())));
	spanModelList.add(new SpanModel("multiple", Color.parseColor("#FFFF5722"), "multiple", null, Typeface.create(Typeface.MONOSPACE, Typeface.BOLD_ITALIC), TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, 13, this.getResources().getDisplayMetrics())));
	spanModelList.add(new SpanModel("words", Color.parseColor("#FF9C27B0"), "words", true, Typeface.create(Typeface.SANS_SERIF, Typeface.BOLD), null));
	new Spantastic.Builder(tvFive, sentence)
                    .setCustomSpanModel(spanModelList)
                    .setClickCallback(new Spantastic.SpannableCallBack() {
                        @Override
                        public void onSpanClick(String spanString, Object... object) {
                            Toast.makeText(MainActivity.this,
                                    String.format(Locale.ENGLISH, "\"%s\" Word Clicked", spanString),
                                    Toast.LENGTH_SHORT).show();
                        }
                    })
                    .apply();


    Kotlin

    sentence = "Spantastic with custom styles for multiple words."
    val tvFive = findViewById<TextView>(R.id.tv_five)
    val spanModelList: MutableList<SpanModel> = ArrayList()
    spanModelList.add(
        SpanModel(
            "custom",
            ContextCompat.getColor(this, R.color.colorPrimary), "custom",
            true, Typeface.create(Typeface.DEFAULT, Typeface.NORMAL),
            TypedValue.applyDimension(
                TypedValue.COMPLEX_UNIT_SP, 15f,
                this.resources.displayMetrics
            )
        )
    )
    spanModelList.add(
        SpanModel(
            "styles",
            ContextCompat.getColor(this, R.color.colorPrimaryDark), "styles",
            false, Typeface.create(Typeface.DEFAULT, Typeface.BOLD),
            TypedValue.applyDimension(
                TypedValue.COMPLEX_UNIT_SP, 10f,
                this.resources.displayMetrics
            )
        )
    )
    spanModelList.add(
        SpanModel(
            "for",
            ContextCompat.getColor(this, R.color.colorAccent), "for",
            true, Typeface.create(Typeface.SERIF, Typeface.ITALIC),
            TypedValue.applyDimension(
                TypedValue.COMPLEX_UNIT_SP, 18f,
                this.resources.displayMetrics
            )
        )
    )
    spanModelList.add(
        SpanModel(
            "multiple",
            Color.parseColor("#FFFF5722"), "multiple", null,
            Typeface.create(Typeface.MONOSPACE, Typeface.BOLD_ITALIC),
            TypedValue.applyDimension(
                TypedValue.COMPLEX_UNIT_SP, 13f,
                this.resources.displayMetrics
            )
        )
    )
    spanModelList.add(
        SpanModel(
            "words",
            Color.parseColor("#FF9C27B0"), "words", true,
            Typeface.create(Typeface.SANS_SERIF, Typeface.BOLD), null
        )
    )
    Spantastic.Builder(tvFive, sentence)
                    .setCustomSpanModel(spanModelList)
                    .setClickCallback(object : Spantastic.SpannableCallBack {
                        override fun onSpanClick(spanString: String?, vararg `object`: Any?) {
                            Toast.makeText(
                                this@KotlinActivity, String.format(
                                    Locale.ENGLISH, "\"%s\" Word Clicked", spanString
                                ),
                                Toast.LENGTH_SHORT
                            ).show()
                        }
                    })

### Above Example's Output

![spantastic_library_custom_spannable_textview](https://user-images.githubusercontent.com/46309117/95648356-09f0b500-0af4-11eb-8f85-e719b3c340e9.png)

## Test Demo

Download DEMO APK: https://github.com/thesarangal/spantasticlibrary/raw/master/SpantasticDemo.apk

## Contributions

Contributions to this library are welcome. If you find a bug or have a feature request,
please open an issue on
the [GitHub repository](https://github.com/thesarangal/spantasticlibrary).

## License

This library is released under the [MIT License](https://opensource.org/licenses/MIT).
You are free to use, modify, and distribute the library for commercial and non-commercial purposes, with attribution to the author.

#### Developed with ❤ by Sarangal


CHANGELOG.md
# Change Log

### Version 2.1.2 (2023-04-02)
- Optimized

### Version 2.1.1 (2022-03-05)
- Optimized

### Version 2.1.0 (2021-06-06)
- Some Bugs Fixed

### Version 2.0.7 (2021-06-02)
- Some Bugs Fixed

### Version 2.0.6 (2021-05-16)
- Kotlin Version
- Some Bugs Fixed

### Version 2.0.2 (2019-12-26)
- Some Bugs Fixed

### Version 2.0.0 (2019-12-13)
- Builder Approach Implemented

### Version 1.0.0 (2019-11-27)
- Publish to jitpack
- Initial version


app/src/test/java/in/sarangal/spannabletextview/ExampleUnitTest.java
package in.sarangal.spannabletextview;

import org.junit.Test;

import static org.junit.Assert.*;

/**
 * Example local unit test, which will execute on the development machine (host).
 *
 * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
 */
public class ExampleUnitTest {
    @Test
    public void addition_isCorrect() {
        assertEquals(4, 2 + 2);
    }
}

spantastic/src/test/java/in/sarangal/lib/spantastic/ExampleUnitTest.java
package in.sarangal.lib.spantastic;

import org.junit.Test;

import static org.junit.Assert.*;

/**
 * Example local unit test, which will execute on the development machine (host).
 *
 * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
 */
public class ExampleUnitTest {
    @Test
    public void addition_isCorrect() {
        assertEquals(4, 2 + 2);
    }
}

spantastic/src/main/java/in/sarangal/lib/spantastic/SpanModel.kt
package `in`.sarangal.lib.spantastic

import android.graphics.Typeface
import androidx.annotation.Keep

/**
 * Developed with Love by Sarangal
 */

/**
 * Class Params:
 *
 * @param textSize     PIXEL value in Float type for TextSize.
 * @param spanString   String which will be convert to spannable.
 * @param colorId      The desired resource identifier, as generated by the aapt
 * tool. This integer encodes the package, type, and resource
 * entry. The value 0 is an invalid identifier.
 * @param callbackKey  The key with return with click on spannable.
 * @param showUnderline true: underline will be display under span text,
 * false: no underline
 * @param typeface     Typeface for custom styling.
 */
@Keep
data class SpanModel(
    val spanString: String,
    val colorId: Int? = null,
    val callbackKey: String,
    val showUnderline: Boolean? = null,
    val typeface: Typeface? = null,
    val textSize: Float? = null
)

app/src/main/java/in/sarangal/spannabletextview/KotlinActivity.kt
package `in`.sarangal.spannabletextview

import `in`.sarangal.lib.spantastic.SpanModel
import `in`.sarangal.lib.spantastic.Spantastic
import android.graphics.Color
import android.graphics.Typeface
import android.os.Bundle
import android.util.TypedValue
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import java.util.*
import kotlin.collections.ArrayList

class KotlinActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        /* Just Show Underline under specific word */
        var sentence = "Spannable with single word underline."
        val tvOne = findViewById<TextView>(R.id.tv_one)
        Spantastic.Builder(tvOne, sentence)
            .setSpan("underline")
            .showUnderline(true)
            .apply()

        /* Show Underline and make Text BOLD for Multiple words */
        sentence = "Spannable with underline and bold for multiple words."
        val strings: ArrayList<String> = ArrayList()
        strings.add("spannable")
        strings.add("underline")
        strings.add("bold")
        val tvTwo = findViewById<TextView>(R.id.tv_two)
        Spantastic.Builder(tvTwo, sentence)
            .setSpanList(strings)
            .showUnderline(true)
            .setTypeface(Typeface.create(Typeface.DEFAULT, Typeface.BOLD))
            .apply()

        /* Set Click on Multiple Words with Bold Style */
        sentence =
            "Spannable with click, color, underline and bold for multiple words."
        strings.clear()
        strings.add("click")
        strings.add("color")
        strings.add("underline")
        strings.add("bold")
        val tvThree = findViewById<TextView>(R.id.tv_three)
        Spantastic.Builder(tvThree, sentence)
            .setSpanList(strings)
            .showUnderline(true)
            .setSpanColor(ContextCompat.getColor(this, R.color.colorAccent))
            .setTypeface(Typeface.create(Typeface.DEFAULT, Typeface.BOLD))
            .setClickCallback(object : Spantastic.SpannableCallBack {
                override fun onSpanClick(spanString: String?, vararg `object`: Any?) {
                    Toast.makeText(
                        this@KotlinActivity, String.format(
                            Locale.ENGLISH,
                            "\"%s\" Word Clicked", spanString
                        ), Toast.LENGTH_SHORT
                    )
                        .show()
                }
            })
            .apply()

        /* Set TextSize */
        sentence = "Spantastic with custom text size for words."
        val tvFour = findViewById<TextView>(R.id.tv_four)
        Spantastic.Builder(tvFour, sentence)
            .setSpan("text size")
            .setTextSize(
                TypedValue.applyDimension(
                    TypedValue.COMPLEX_UNIT_SP, 20f,
                    this.resources.displayMetrics
                )
            )
            .apply()

        /* Set Custom Spannables with unique styling */
        sentence =
            "Spantastic with custom styles for multiple words."
        val tvFive = findViewById<TextView>(R.id.tv_five)
        val spanModelList: MutableList<SpanModel> = ArrayList()
        spanModelList.add(
            SpanModel(
                "custom",
                ContextCompat.getColor(this, R.color.colorPrimary), "custom",
                true, Typeface.create(Typeface.DEFAULT, Typeface.NORMAL),
                TypedValue.applyDimension(
                    TypedValue.COMPLEX_UNIT_SP, 15f,
                    this.resources.displayMetrics
                )
            )
        )
        spanModelList.add(
            SpanModel(
                "styles",
                ContextCompat.getColor(this, R.color.colorPrimaryDark), "styles",
                false, Typeface.create(Typeface.DEFAULT, Typeface.BOLD),
                TypedValue.applyDimension(
                    TypedValue.COMPLEX_UNIT_SP, 10f,
                    this.resources.displayMetrics
                )
            )
        )
        spanModelList.add(
            SpanModel(
                "for",
                ContextCompat.getColor(this, R.color.colorAccent), "for",
                true, Typeface.create(Typeface.SERIF, Typeface.ITALIC),
                TypedValue.applyDimension(
                    TypedValue.COMPLEX_UNIT_SP, 18f,
                    this.resources.displayMetrics
                )
            )
        )
        spanModelList.add(
            SpanModel(
                "multiple",
                Color.parseColor("#FFFF5722"), "multiple", null,
                Typeface.create(Typeface.MONOSPACE, Typeface.BOLD_ITALIC),
                TypedValue.applyDimension(
                    TypedValue.COMPLEX_UNIT_SP, 13f,
                    this.resources.displayMetrics
                )
            )
        )
        spanModelList.add(
            SpanModel(
                "words",
                Color.parseColor("#FF9C27B0"), "words", true,
                Typeface.create(Typeface.SANS_SERIF, Typeface.BOLD), null
            )
        )
        Spantastic.Builder(tvFive, sentence)
            .setCustomSpanModel(spanModelList)
            .setClickCallback(object : Spantastic.SpannableCallBack {
                override fun onSpanClick(spanString: String?, vararg `object`: Any?) {
                    Toast.makeText(
                        this@KotlinActivity, String.format(
                            Locale.ENGLISH, "\"%s\" Word Clicked", spanString
                        ),
                        Toast.LENGTH_SHORT
                    ).show()
                }
            })
            .apply()

        /* Developed By Sarangal */
        sentence = "Developed with ❤ by Sarangal"
        val tvCredit = findViewById<TextView>(R.id.tv_credit)
        val creditModelList = ArrayList<SpanModel>()
        creditModelList.add(
            SpanModel(
                "❤", Color.parseColor("#FFFF2F2F"),
                "Love", null,
                Typeface.create(Typeface.DEFAULT, Typeface.NORMAL), null
            )
        )
        creditModelList.add(
            SpanModel(
                "Sarangal",
                Color.parseColor("#FF00BCD4"), "Sarangal", null,
                Typeface.create(Typeface.DEFAULT, Typeface.BOLD), null
            )
        )
        Spantastic.Builder(tvCredit, sentence)
            .setCustomSpanModel(creditModelList)
            .apply()
    }
}

spantastic/src/main/java/in/sarangal/lib/spantastic/Spantastic.kt
package `in`.sarangal.lib.spantastic

import android.graphics.Color
import android.graphics.Typeface
import android.text.SpannableString
import android.text.TextPaint
import android.text.method.LinkMovementMethod
import android.text.style.ClickableSpan
import android.view.View
import android.widget.TextView
import java.util.*
import java.util.regex.Pattern

/**
 * Developed with Love by Sarangal
 */
class Spantastic private constructor(builder: Builder) {

    /*
    * Data Members
    * */
    private val textView: TextView = builder.textView
    private val fullString: String = builder.completeString
    private val colorID: Int? = builder.colorID
    private val spannableCallBack: SpannableCallBack? = builder.spannableCallBack
    private val spanModelList: List<SpanModel>? = builder.spanModelList
    private val typeface: Typeface? = builder.typeface
    private val textSize: Float? = builder.textSize
    private val showUnderline: Boolean = builder.showUnderline
    private val objectAny: Any? = builder.objectAny

    init {

        /* Set Styling to the TextView */
        setSpannable()
    }

    /**
     * Set Styling to the TextView
     * */
    private fun setSpannable() {

        if (fullString.isBlank()) {
            return
        }

        if (spanModelList.isNullOrEmpty()) {
            return
        }

        val spannableString = SpannableString(fullString)

        /* Iterate Span List */
        spanModelList.forEach { spanModel ->

            if (spanModel.spanString.isEmpty()) {
                return@forEach
            }

            var spanString = spanModel.spanString

            try {
                /* Check DOLLAR ($) Sign in String */
                if (spanString.contains("$")) {
                    spanString = spanString.replace("$", "\\$")
                }

                val pattern = Pattern.compile(spanString)
                val matcher = pattern.matcher(fullString)
                while (matcher.find()) {
                    val start = matcher.start()
                    val end = matcher.end()
                    val clickSpan: ClickableSpan = object : ClickableSpan() {
                        override fun onClick(view: View) {
                            spannableCallBack?.onSpanClick(
                                spanModel.callbackKey, objectAny
                            )
                        }

                        override fun updateDrawState(textPaint: TextPaint) {
                            super.updateDrawState(textPaint)

                            /* Set Span UnderLine */
                            textPaint.isUnderlineText = spanModel.showUnderline ?: showUnderline

                            /* Set Span Color */
                            try {
                                textPaint.color = spanModel.colorId ?: colorID
                                        ?: textView.currentTextColor
                            } catch (e: Exception) {
                                e.printStackTrace()
                            }

                            /* Set Span Typeface for Styling */
                            (spanModel.typeface ?: typeface)?.let {
                                textPaint.typeface = it
                            }

                            /* Set Span TextSize */
                            (spanModel.textSize ?: textSize)?.let {
                                textPaint.textSize = it
                            }
                        }
                    }

                    /* Set Span on SpannableString */
                    spannableString.setSpan(clickSpan, start, end, 0)
                }

            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        /* Set Spannable to TextView */
        textView.text = spannableString

        /* Support to Clickable */
        textView.movementMethod = LinkMovementMethod.getInstance()
        textView.highlightColor = Color.TRANSPARENT
    }

    /**
     * Interface for Spannable Click CallBack
     */
    interface SpannableCallBack {
        fun onSpanClick(spanString: String?, vararg `object`: Any?)
    }

    /**
     * Builder Class for Handling Spannable Operations
     *
     * Method Params:
     *
     * @param textView   TextView on which spannable text will be applied
     * @param completeString Complete string value from which span(s) will be found.
     */
    open class Builder(internal val textView: TextView, internal val completeString: String) {
        internal var colorID: Int? = null
        internal var spannableCallBack: SpannableCallBack? = null
        internal var spanModelList: List<SpanModel>? = null
        internal var typeface: Typeface? = null
        internal var showUnderline = false
        internal var textSize: Float? = null
        internal var objectAny: Any? = null

        /**
         * Method Params:
         *
         * @param singleString Single string which will be convert to spannable.
         */
        fun setSpan(singleString: String): Builder {
            val spanList: MutableList<String> = ArrayList()
            if (singleString.isNotEmpty()) {
                spanList.add(singleString)
            }
            return setSpanList(spanList)
        }

        /**
         * Method Params:
         *
         * @param spanList List of strings which will be convert to spannable.
         */
        fun setSpanList(spanList: List<String>): Builder {
            val spanModelList: MutableList<SpanModel> = ArrayList()

            /* Iterate Span List */
            spanList.forEach { spanString ->

                if (spanString.isEmpty()) {
                    return@forEach
                }

                spanModelList.add(SpanModel(spanString, callbackKey = spanString))
            }

            return setCustomSpanModel(spanModelList)
        }

        /**
         * Method Params:
         *
         * @param colorId The desired resource identifier, as generated by the aapt
         * tool. This integer encodes the package, type, and resource
         * entry. The value 0 is an invalid identifier.
         */
        fun setSpanColor(colorId: Int): Builder {
            colorID = colorId
            return this
        }

        /**
         * Method Params:
         *
         * @param showUnderline true: underline will be display under span text,
         */
        fun showUnderline(showUnderline: Boolean): Builder {
            this.showUnderline = showUnderline
            return this
        }

        /**
         * Method Params:
         *
         * @param spannableCallBack Interface reference to get click callback of specific
         */
        fun setClickCallback(spannableCallBack: SpannableCallBack): Builder {
            this.spannableCallBack = spannableCallBack
            return this
        }

        /**
         * Method Params:
         *
         * @param typeface Typeface for custom styling.
         */
        fun setTypeface(typeface: Typeface): Builder {
            this.typeface = typeface
            return this
        }

        /**
         * Method Params:
         *
         * @param textSize PIXEL value in Float type for TextSize.
         */
        fun setTextSize(textSize: Float): Builder {
            this.textSize = textSize
            return this
        }

        /**
         * Method Params:
         *
         * @param spanModelList List of custom model of strings with custom
         * styling params which will be convert to spannable.
         */
        fun setCustomSpanModel(spanModelList: List<SpanModel>): Builder {
            this.spanModelList = spanModelList
            return this
        }

        /**
         * Method Params:
         *
         * @param object May object any object which will be return on span click via
         * spannableCallBack interface.
         */
        fun setObject(`object`: Any): Builder {
            objectAny = `object`
            return this
        }

        /**
         * Implement Spantastic Constructor method.
         */
        fun apply() {
            Spantastic(this)
        }
    }
}

app/src/main/java/in/sarangal/spannabletextview/MainActivity.java
package in.sarangal.spannabletextview;

import android.graphics.Color;
import android.graphics.Typeface;
import android.os.Bundle;
import android.util.TypedValue;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import in.sarangal.lib.spantastic.SpanModel;
import in.sarangal.lib.spantastic.Spantastic;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        String sentence;
        List<String> strings;

        /* Just Show Underline under specific word */
        sentence = "Spannable with single word underline.";
        TextView tvOne = findViewById(R.id.tv_one);
        new Spantastic.Builder(tvOne, sentence)
                .setSpan("underline")
                .showUnderline(true)
                .apply();

        /* Show Underline and make Text BOLD for Multiple words */
        sentence = "Spannable with underline and bold for multiple words.";
        strings = new ArrayList<>();
        strings.add("spannable");
        strings.add("underline");
        strings.add("bold");
        TextView tvTwo = findViewById(R.id.tv_two);
        new Spantastic.Builder(tvTwo, sentence)
                .setSpanList(strings)
                .showUnderline(true)
                .setTypeface(Typeface.create(Typeface.DEFAULT, Typeface.BOLD))
                .apply();

        /* Set Click on Multiple Words with Bold Style */
        sentence = "Spannable with click, color, underline and bold for multiple words.";
        strings = new ArrayList<>();
        strings.add("click");
        strings.add("color");
        strings.add("underline");
        strings.add("bold");
        TextView tvThree = findViewById(R.id.tv_three);
        new Spantastic.Builder(tvThree, sentence)
                .setSpanList(strings)
                .showUnderline(true)
                .setSpanColor(ContextCompat.getColor(this, R.color.colorAccent))
                .setTypeface(Typeface.create(Typeface.DEFAULT, Typeface.BOLD))
                .setClickCallback((spanString, object) -> Toast.makeText(MainActivity.this, String.format(Locale.ENGLISH,
                        "\"%s\" Word Clicked", spanString), Toast.LENGTH_SHORT)
                        .show())
                .apply();

        /* Set TextSize */
        sentence = "Spantastic with custom text size for words.";
        TextView tvFour = findViewById(R.id.tv_four);
        new Spantastic.Builder(tvFour, sentence)
                .setSpan("text size")
                .setTextSize(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, 20,
                        this.getResources().getDisplayMetrics()))
                .apply();

        /* Set Custom Spannables with unique styling */
        sentence = "Spantastic with custom styles for multiple words.";
        TextView tvFive = findViewById(R.id.tv_five);
        List<SpanModel> spanModelList = new ArrayList<>();
        spanModelList.add(new SpanModel("custom",
                ContextCompat.getColor(this, R.color.colorPrimary), "custom",
                true, Typeface.create(Typeface.DEFAULT, Typeface.NORMAL),
                TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, 15,
                        this.getResources().getDisplayMetrics())));
        spanModelList.add(new SpanModel("styles",
                ContextCompat.getColor(this, R.color.colorPrimaryDark), "styles",
                false, Typeface.create(Typeface.DEFAULT, Typeface.BOLD),
                TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, 10,
                        this.getResources().getDisplayMetrics())));
        spanModelList.add(new SpanModel("for",
                ContextCompat.getColor(this, R.color.colorAccent), "for",
                true, Typeface.create(Typeface.SERIF, Typeface.ITALIC),
                TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, 18,
                        this.getResources().getDisplayMetrics())));
        spanModelList.add(new SpanModel("multiple",
                Color.parseColor("#FFFF5722"), "multiple", null,
                Typeface.create(Typeface.MONOSPACE, Typeface.BOLD_ITALIC),
                TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, 13,
                        this.getResources().getDisplayMetrics())));
        spanModelList.add(new SpanModel("words",
                Color.parseColor("#FF9C27B0"), "words", true,
                Typeface.create(Typeface.SANS_SERIF, Typeface.BOLD), null));
        new Spantastic.Builder(tvFive, sentence)
                .setCustomSpanModel(spanModelList)
                .setClickCallback((spanString, object) -> Toast.makeText(MainActivity.this,
                        String.format(Locale.ENGLISH, "\"%s\" Word Clicked", spanString),
                        Toast.LENGTH_SHORT).show())
                .apply();

        /* Developed By Sarangal */
        sentence = "Developed with ❤ by Sarangal";
        TextView tvCredit = findViewById(R.id.tv_credit);
        List<SpanModel> creditModelList = new ArrayList<>();
        creditModelList.add(new SpanModel("❤", Color.parseColor("#FFFF2F2F"),
                "Love", null,
                Typeface.create(Typeface.DEFAULT, Typeface.NORMAL), null));
        creditModelList.add(new SpanModel("Sarangal",
                Color.parseColor("#FF00BCD4"), "Sarangal", null,
                Typeface.create(Typeface.DEFAULT, Typeface.BOLD), null));
        new Spantastic.Builder(tvCredit, sentence)
                .setCustomSpanModel(creditModelList)
                .apply();


    }
}
